lib/src/service/internal_key_event_handlers/copy_paste_handler.dart

```diff
@@ -73,3 +82,217 @@ void handlePastePlainText(EditorState editorState, String plainText) {
     _pasteMarkdown(editorState, plainText);
   }
 }
+
+void pasteHTML(EditorState editorState, String html) {
+  final selection = editorState.selection?.normalized;
+  if (selection == null) {
+    return;
+  }
+
+  assert(selection.isCollapsed);
+
+  Log.keyboard.debug('paste html: $html');
+  final htmltoNodes = htmlToDocument(html);
+
+  if (htmltoNodes.isEmpty) {
+    return;
+  }
+
+  _pasteMultipleLinesInText(
+    editorState,
+    selection.start.offset,
+    htmltoNodes.root.children.toList(),
+  );
+}
+
+Selection _computeSelectionAfterPasteMultipleNodes(
+  EditorState editorState,
+  List<Node> nodes,
+) {
+  final currentSelection = editorState.selection!;
+  final currentCursor = currentSelection.start;
+  final currentPath = [...currentCursor.path];
+  currentPath[currentPath.length - 1] += nodes.length;
+  int lenOfLastNode = _textLengthOfNode(nodes.last);
+  return Selection.collapsed(
+    Position(path: currentPath, offset: lenOfLastNode),
+  );
+}
+
+void handleCopy(EditorState editorState) async {
+  final selection = editorState.selection?.normalized;
+  if (selection == null || selection.isCollapsed) {
+    return;
+  }
+  final text = editorState.getTextInSelection(selection).join('\n');
+  final nodes = editorState.getSelectedNodes(selection);
+  if (nodes.isEmpty) {
+    return;
+  }
+  final html = documentToHTML(
+    Document(
+      root: Node(
+        type: 'page',
+        children: nodes,
+      ),
+    ),
+  );
+  AppFlowyClipboard.setData(
+    text: text,
+    html: html.isEmpty ? null : html,
+  );
+  return;
+}
+
+void _pasteMultipleLinesInText(
+  EditorState editorState,
+  int offset,
+  List<Node> nodes,
+) {
+  final tb = editorState.transaction;
+  final selection = editorState.selection;
+  final afterSelection =
+      _computeSelectionAfterPasteMultipleNodes(editorState, nodes);
+
+  final selectionNode = editorState.getNodesInSelection(selection!);
+  if (selectionNode.length == 1) {
+    final node = selectionNode.first;
+    final delta = node.delta;
+    if (delta == null) {
+      tb.afterSelection = afterSelection;
+      tb.insertNodes(afterSelection.end.path, nodes);
+      editorState.apply(tb);
+    }
```

Comment:
Seems you don't need delta variable much.

```suggestion
    final node = selectionNode.first;
    if (node.delta == null) {
      tb.afterSelection = afterSelection;
      tb.insertNodes(afterSelection.end.path, nodes);
      editorState.apply(tb);
    }
```